### PR TITLE
refactor: use TLS library v4

### PR DESCRIPTION
**Pull request by gruyaume**
# Description

Bump the TLS library to V4.

## Addition of certificate helper methods

The addition of `src/certificates.py` is because the TLS library `V4` does not provide the `generate_certificate`, `generate_private_key`, and `generate_ca` helpers as those were specific to this self-signed certificates charm. 

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: tests/unit/certificates_helpers.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def generate_private_key(
+    key_size: int = 2048,
+    public_exponent: int = 65537,
+) -> str:
+    """Generate a private key.
+
+    Args:
+        password (bytes): Password for decrypting the private key
```
Comment (saltiyazan) [inline tests/unit/certificates_helpers.py:20]: This isn't there anymore.
```suggestion
```

File: src/certificates.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption(),
+    )
+    return key_bytes.decode().strip()
+
+
+def get_certificate_extensions(
+    authority_key_identifier: bytes,
```
Comment (saltiyazan) [inline src/certificates.py:46]: More of a nitpicking but I this could be optional too.

File: src/certificates.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    )
+    key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.TraditionalOpenSSL,
+        encryption_algorithm=serialization.NoEncryption(),
+    )
+    return key_bytes.decode().strip()
+
+
+def get_certificate_extensions(
```
Comment (saltiyazan) [inline src/certificates.py:45]:
```suggestion
def get_certificate_request_extensions(
```

File: src/certificates.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+
+    cert = certificate_builder.sign(private_key, hashes.SHA256())  # type: ignore[arg-type]
+    return cert.public_bytes(serialization.Encoding.PEM).decode().strip()
+
+
+def generate_ca(
+    private_key: str,
+    subject: str,
+    validity: int = 365,
+    country: str = "US",
```
Comment (saltiyazan) [inline src/certificates.py:205]: Are the defaults necessary? If I'm not mistaken the country is not mandatory and the validity is as important as the subject for the caller to specify.

File: src/charm.py [hunk 5]
```diff
@@ -158,17 +154,14 @@ def _generate_root_certificate(self) -> None:
         """
         if not self._config_ca_common_name:
             raise ValueError("CA common name should not be empty")
-        private_key_password = generate_password()
-        private_key = generate_private_key(password=private_key_password.encode())
+        private_key = generate_private_key()
         ca_certificate = generate_ca(
             private_key=private_key,
             subject=self._config_ca_common_name,
```
Comment (saltiyazan) [inline src/charm.py:160]: I think it's worth adding the validity here and not rely on the default value.
It should come from `_config_root_ca_certificate_validity`

File: src/charm.py [hunk 6]
```diff
@@ -158,17 +154,14 @@ def _generate_root_certificate(self) -> None:
...
         )
         secret_content = {
-            "private-key-password": private_key_password,
-            "private-key": private_key.decode(),
-            "ca-certificate": ca_certificate.decode(),
+            "private-key": private_key,
+            "ca-certificate": ca_certificate,
         }
         if self._root_certificate_is_stored:
             secret = self.model.get_secret(label=CA_CERTIFICATES_SECRET_LABEL)
```
Comment (saltiyazan) [inline src/charm.py:167]: I know this wasn't done in this PR but the CA secret should expire based on `_config_root_ca_certificate_validity` not on `_config_certificate_validity`

File: src/certificates.py (hunk 4 above)
Comment (saltiyazan) [reply src/certificates.py:205]: `Country` should probably come from user config if we want to keep it at all.

File: tests/unit/test_certificates.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    certificate = generate_certificate(csr=csr, ca=ca, ca_key=ca_key)
+
+    cert = x509.load_pem_x509_certificate(certificate.encode())
+    result_all_sans = cert.extensions.get_extension_for_class(x509.SubjectAlternativeName)
+
+    result_sans_dns = sorted(result_all_sans.value.get_values_for_type(x509.DNSName))
+    assert result_sans_dns == sorted(set(sans))
+
+
+def test_given_basic_constraint_is_false_when_generate_ca_then_extensions_are_correctly_populated():  # noqa: E501
```
Comment (saltiyazan) [inline tests/unit/test_certificates.py:145]: What does it mean for the basic constraint to be false?

File: tests/unit/test_certificates.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    public_key.verify(  # type: ignore[call-arg, union-attr]
+        certificate_object.signature,
+        certificate_object.tbs_certificate_bytes,
+        padding.PKCS1v15(),  # type: ignore[arg-type]
+        certificate_object.signature_hash_algorithm,  # type: ignore[arg-type]
+    )
+
+
+def test_given_request_is_for_ca_certificate_when_generate_certificate_then_certificate_is_generated():  # noqa: E501
+    ca_private_key = generate_private_key()
```
Comment (saltiyazan) [inline tests/unit/test_certificates.py:194]: why not using the helper function?

**Review by saltiyazan — CHANGES_REQUESTED** [review]
I went through it and it looks really good. I have a few comments some of which might not be directly on the PR but I think they could be included here.

File: src/certificates.py
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/certificates.py:1]: Which other TLS provider would generate certs using these methods? They were really for the self-signed certificates charm.

This change was explained here:
- https://github.com/canonical/tls-certificates-interface/pull/214

File: src/certificates.py (hunk 2 above)
Comment (gruyaume) [reply src/certificates.py:46]: Why? The only method that calls `get_certificate_extensions` uses this parameter

File: tests/unit/certificates_helpers.py (hunk 1 above)
Comment (gruyaume) [reply tests/unit/certificates_helpers.py]: Done

File: src/certificates.py (hunk 4 above)
Comment (gruyaume) [reply src/certificates.py:205]: While I do agree with those comments, here I simply took what was in the lib v3 and moved it to self-signed certificates. Making the country a Juju config makes sense but I would see this as a separate task.

File: src/charm.py (hunk 6 above)
Comment (gruyaume) [reply src/charm.py:167]: I agree but this is unrelated to this change, I suggest keeping this task focused on this refactoring and not include other bug fixes

File: src/charm.py (hunk 5 above)
Comment (gruyaume) [reply src/charm.py:160]: I agree but this is unrelated to this change right? I would prefer to do it in a separate PR

File: tests/unit/test_certificates.py (hunk 8 above)
Comment (gruyaume) [reply tests/unit/test_certificates.py]: Good catch, done

File: tests/unit/test_certificates.py (hunk 7 above)
Comment (gruyaume) [reply tests/unit/test_certificates.py]: I improved the test name, ultimately it's more that when we generate a certificate using generate_ca, we expect the BasicConstraint extension to be set

File: src/charm.py [hunk 9]
```diff
@@ -214,8 +207,8 @@ def _process_outstanding_certificate_requests(self) -> None:
         """Process outstanding certificate requests."""
         for request in self.tls_certificates.get_outstanding_certificate_requests():
             self._generate_self_signed_certificate(
-                csr=request.csr,
-                is_ca=request.is_ca,
+                csr=str(request.certificate_signing_request),
```
Comment (kayra1) [inline src/charm.py:210]: why not `request.certificate_signing_request.raw`?

**Review by kayra1 — APPROVED** [review]
one small comment

File: src/certificates.py
[the remark this answers is not in the document]
Comment (saltiyazan) [reply src/certificates.py:1]: Not necessarily the providers we maintain. I think they were useful helpers in the lib. But I'm fine with it especially that it has been agreed upon in the lib PR.

File: src/certificates.py (hunk 3 above)
Comment (saltiyazan) [reply src/certificates.py:45]: Aren't they? I mean the current name `get_certificate_extensions` would suggest that when implementing some feature I can use this function to see what extensions does the certificate have.
The docstring makes it clearer that that isn't the intended usage.
But the way this function is being used is that it is extracting extensions from the csr (cert request) and from parameters that come from the requirer's certificate request.
So at the end we are getting extensions from the request that we want to add to the certificate.

It's not a big deal, but the current name is misleading for me.
how about something like `get_certificate_extensions`?

File: src/certificates.py (hunk 4 above)
Comment (saltiyazan) [reply src/certificates.py:197]: Sure that makes sense. Let's do it in a separate task.

Comment (saltiyazan) [reply src/certificates.py:197]: Ticket created.

File: src/charm.py (hunk 5 above)
Comment (saltiyazan) [reply src/charm.py:160]: Issue created. #213

File: src/charm.py (hunk 6 above)
Comment (saltiyazan) [reply src/charm.py:167]: Ticket created #214

**Review by saltiyazan — APPROVED** [review]
Looks good. I created tickets for the comments that would not be addressed as part of this PR.
For the comment on `get_certificate_extensions` I prefer a better name, but I'll leave it up to you to resolve the comment either accepting or rejecting the new suggestion.

File: src/certificates.py
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/certificates.py:1]: I'm fine adding those back in the lib if we find they would be useful to more than this charm

File: src/certificates.py (hunk 3 above)
Comment (gruyaume) [reply src/certificates.py]: Sorry I had answered too quick there, I used your recommendation.

File: src/charm.py (hunk 9 above)
Comment (gruyaume) [reply src/charm.py:210]: I mean, those two are equivalent, I just find that str(abc) looked cleaner than abc.raw. I don't really care either way.

**Pull request merged**